coordinator/history/history.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package history
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	hashSize = 32 // byte, History.hashFun().Size()
+	histPath = "/mnt/state/history"
+)
+
+// History is the history of the Coordinator.
+type History struct {
+	store      store
+	hashFun    func() hash.Hash
+	signingKey *ecdsa.PrivateKey
+}
+
+// New creates a new History with the given signing key.
+func New() *History {
+	osFS := afero.NewOsFs()
+	if err := osFS.MkdirAll(histPath, 0o755); err != nil {
+		panic(fmt.Errorf("creating history directory: %w", err))
+	}
+	h := &History{
+		store:   newPVStore(&afero.Afero{Fs: afero.NewBasePathFs(osFS, histPath)}),
+		hashFun: sha256.New,
+	}
+	if hashSize != h.hashFun().Size() {
+		panic("mismatch between hashSize and hash function size")
+	}
+	return h
+}
+
+// ConfigureSigningKey sets the signing key for validation and signing of the protected history parts.
+func (h *History) ConfigureSigningKey(signingKey *ecdsa.PrivateKey) {
+	h.signingKey = signingKey
+}
+
+// GetManifest returns the manifest for the given hash.
+func (h *History) GetManifest(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed(hash, "manifests/%s")
+}
+
+// SetManifest sets the manifest and returns its hash.
+func (h *History) SetManifest(manifest []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("manifests/%s", manifest)
+}
+
+// GetPolicy returns the policy for the given hash.
+func (h *History) GetPolicy(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed(hash, "policies/%s")
+}
+
+// SetPolicy sets the policy and returns its hash.
+func (h *History) SetPolicy(policy []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("policies/%s", policy)
+}
+
+// GetTransition returns the transition for the given hash.
+func (h *History) GetTransition(hash [hashSize]byte) (Transition, error) {
+	transitionBytes, err := h.getContentaddressed(hash, "transitions/%s")
+	if err != nil {
+		return Transition{}, err
+	}
+	var transition Transition
+	if err := transition.unmarshalBinary(transitionBytes); err != nil {
+		return Transition{}, fmt.Errorf("unmarshaling transition: %w", err)
+	}
+	return transition, nil
+}
+
+// SetTransition sets the transition and returns its hash.
+func (h *History) SetTransition(transition Transition) ([hashSize]byte, error) {
+	return h.setContentaddressed("transitions/%s", transition.marshalBinary())
+}
+
+// GetLatest returns the verified transition for the given hash.
+func (h *History) GetLatest() (LatestTransition, error) {
+	if h.signingKey == nil {
+		return LatestTransition{}, errors.New("signing key not configured")
+	}
+	transitionBytes, err := h.store.Get("transitions/latest")
```

Comment:
I believe we forgot to specify that `GetLatest` needs to respect the mutual exclusion properties of `SetLatest`, so that we don't end up reading a half-written file :/

Reply:
That should be given by the store's mutex, right?

Reply:
We could just require the store interface to behave well on concurrent Set and Get, and implement Set by writing to a random file and then atomically `syscall.Rename` it.

Reply:
> That should be given by the store's mutex, right?

Sorry, missed that.